### PR TITLE
Port display improvements

### PR DESCRIPTION
**Pull request by wdanilo**
This PR redefines how ports are being drawn and handled. It is a big node port rewrite. The previous implementation was just proof of concept to gather requirements, now we are doing it the right way. This PR should solve the following issues:

1. [Connections on nodes may get misplaced](https://github.com/enso-org/ide/issues/860)
2. [Incorrect connection for a node with a method](https://github.com/enso-org/ide/issues/876)
3. [Connected port Highlight](https://github.com/enso-org/ide/issues/524)
4. [The ports & connections are not colored according to type information](https://github.com/enso-org/ide/issues/765)
5. [Incorrect port display on operator expressions](https://github.com/enso-org/ide/issues/517)
6. [Moving node bugs disabled node](https://github.com/enso-org/ide/issues/869)
7. [Editable expression contains non-connected port names]
8. [Improper edge disconnect highlight color]

### CODE REVIEW

File: src/rust/ensogl/lib/core/src/gui/cursor.rs [hunk 1]
```diff
@@ -335,7 +335,7 @@ impl Cursor {
         let input   = &model.frp;
         let mouse   = &scene.mouse.frp;
 
-        // === Animations ===
+        // === DEPRECATED_Animations ===
```
Comment (farmaazon) [inline src/rust/ensogl/lib/core/src/gui/cursor.rs:338]: Revert :)

File: src/rust/ensogl/lib/theme/src/lib.rs [hunk 2]
```diff
@@ -9,394 +9,326 @@
...
 }
 
-define_theme! { dark
+
+
+// =============================
+// === Light Theme & Modules ===
+// =============================
+
+define_default_theme! { light_theme
```
Comment (farmaazon) [inline src/rust/ensogl/lib/theme/src/lib.rs:104]: let's put it into `theme` module.

File: src/rust/ide/lib/span-tree/src/lib.rs [hunk 3]
```diff
@@ -97,27 +97,51 @@ pub struct SpanTree<T=()> {
...
-        node::Ref {
-            node       : &self.root,
-            span_begin : default(),
-            crumbs     : default(),
-            ast_crumbs : default()
-        }
+        node::Ref::new(&self.root)
+    }
+
+    /// Get the `NodeRef` of root node.
```
Comment (farmaazon) [inline src/rust/ide/lib/span-tree/src/lib.rs:109]: update docs here and in `root_ref`.

File: src/rust/ide/lib/span-tree/src/lib.rs [hunk 4]
```diff
@@ -97,27 +97,51 @@ pub struct SpanTree<T=()> {
...
+    pub fn map<S>(self, f:impl Copy+Fn(T)->S) -> SpanTree<S> {
+        let root = self.root.map(f);
+        SpanTree {root}
+    }
+}
+
+
+// === Getters ===
+
+impl <T:Payload> SpanTree<T> {
```
Comment (farmaazon) [inline src/rust/ide/lib/span-tree/src/lib.rs:130]: spacing

File: src/rust/ide/lib/span-tree/src/node.rs [hunk 5]
```diff
@@ -58,30 +58,44 @@ impl<T:Payload> Node<T> {
...
-    pub fn is_insertion_point(&self) -> bool {
-        self.kind.is_insertion_point()
+    /// Payload mapping utility.
+    pub fn map<S>(self, f:impl Copy+Fn(T)->S) -> Node<S> {
+        let kind     = self.kind;
+        let size     = self.size;
+        let children = self.children.into_iter().map(|t|t.map(f)).collect_vec();
+        let ast_id   = self.ast_id;
+        let payload  = f(self.payload);
+        Node {kind,size,children,ast_id,payload}
```
Comment (farmaazon) [inline src/rust/ide/lib/span-tree/src/node.rs:74]:
```suggestion
        Node {payload,..self}
```

File: src/rust/ide/lib/span-tree/src/node.rs [hunk 6]
```diff
@@ -58,30 +58,44 @@ impl<T:Payload> Node<T> {
...
     }
 }
 
+// === Kind utils ===
+
+#[allow(missing_docs)]
+impl<T:Payload> Node<T> {
+    pub fn is_parensed(&self) -> bool {
```
Comment (farmaazon) [inline src/rust/ide/lib/span-tree/src/node.rs:82]: Add docs

File: src/rust/ide/lib/span-tree/src/node.rs [hunk 7]
```diff
@@ -123,6 +137,16 @@ pub struct Child<T=()> {
     pub ast_crumbs : ast::Crumbs,
 }
 
+impl<T:Payload> Child<T> {
+    /// Payload mapping utility.
+    pub fn map<S>(self, f:impl Copy+Fn(T)->S) -> Child<S> {
+        let node       = self.node.map(f);
+        let offset     = self.offset;
+        let ast_crumbs = self.ast_crumbs;
+        Child {node,offset,ast_crumbs}
```
Comment (farmaazon) [inline src/rust/ide/lib/span-tree/src/node.rs:146]:
```suggestion
        Child {node,..self}
```

File: src/rust/ide/lib/span-tree/src/node.rs [hunk 8]
```diff
@@ -295,26 +382,37 @@ pub struct NodeFoundByAstCrumbs<'a,'b,T=()> {
 }
 
 impl<'a,T:Payload> Ref<'a,T> {
+    /// Constructor.
+    pub fn new(node:&'a Node<T>) -> Self {
+        let span_begin = default();
+        let crumbs     = default();
+        let ast_crumbs = default();
+        Self {node,span_begin,crumbs,ast_crumbs}
```
Comment (farmaazon) [inline src/rust/ide/lib/span-tree/src/node.rs:390]: consider macro "fill_default"

File: src/rust/ide/view/graph-editor/src/component/edge.rs [hunk 9]
```diff
@@ -1319,8 +1318,8 @@ impl EdgeModelData {
 
         // FIXME : StyleWatch is unsuitable here, as it was designed as an internal tool for shape system (#795)
         let styles                        = StyleWatch::new(&self.scene.style_sheet);
-        let split_color_lightness_factor  = styles.get_number_or(theme::vars::graph_editor::edge::split_color::lightness_factor,1.2);
-        let split_color_chroma_factor     = styles.get_number_or(theme::vars::graph_editor::edge::split_color::chroma_factor,0.8);
+        let split_color_lightness_factor  = styles.get_number_or(theme::graph_editor::edge::split_color::lightness_factor,1.2);
+        let split_color_chroma_factor     = styles.get_number_or(theme::graph_editor::edge::split_color::chroma_factor,0.8);
```
Comment (farmaazon) [inline src/rust/ide/view/graph-editor/src/component/edge.rs:1322]: over 100 lines.

File: src/rust/ide/view/graph-editor/src/component/node.rs [hunk 10]
```diff
@@ -55,11 +55,11 @@ pub mod shape {
...
     ensogl::define_shape_system! {
         (style:Style, selection:f32, bg_color:Vector4 ) {
-            use ensogl_theme::vars::graph_editor::node as node_theme;
+            use ensogl_theme::graph_editor::node as node_theme;
 
             let bg_color        = Var::<color::Rgba>::from(bg_color);
-            let selection_color = style.get_color(ensogl_theme::vars::graph_editor::node::selection::color);
-            let _selection_size = style.get_number_or(ensogl_theme::vars::graph_editor::node::selection::size,8.0);
+            let selection_color = style.get_color(ensogl_theme::graph_editor::node::selection::color);
+            let _selection_size = style.get_number_or(ensogl_theme::graph_editor::node::selection::size,8.0);
```
Comment (farmaazon) [inline src/rust/ide/view/graph-editor/src/component/node.rs:62]: Why we declare it? Might be removed.

File: src/rust/ide/view/graph-editor/src/component/node/expression.rs [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    pub output_span_tree : SpanTree,
+}
+
+impl Expression {
+    pub fn debug_from_str(s:&str) -> Self {
+        let code             = s.into();
+        let input_span_tree  = default();
+        let output_span_tree = default();
+        Self {code,input_span_tree,output_span_tree}
+    }
```
Comment (farmaazon) [inline src/rust/ide/view/graph-editor/src/component/node/expression.rs:24]: missing docs.

File: src/rust/ensogl/example/src/list_view.rs [hunk 12]
```diff
@@ -95,8 +95,8 @@ impl list_view::entry::ModelProvider for MockEntries {
 // ========================
 
 fn init(app:&Application) {
-    ensogl_theme::dark::setup(&app);
-    ensogl_theme::light::setup(&app);
+    ensogl_theme::dark_theme::setup(&app);
+    ensogl_theme::light_theme::setup(&app);
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply src/rust/ensogl/example/src/list_view.rs:99]: Changed to `theme::builtin::light`

File: src/rust/ensogl/lib/core/src/application/command.rs [hunk 13]
```diff
@@ -82,20 +82,20 @@ pub trait FrpNetworkProvider {
 #[derive(Debug)]
 #[allow(missing_docs)]
 pub struct Command {
-    pub frp     : frp::Source,
+    pub frp     : frp::Any,
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply src/rust/ensogl/lib/core/src/application/command.rs:85]: I added short notice to `Any` indicating that the detailed docs are attached to `any_mut`. The naming needs to be improved in the future.

File: src/rust/ensogl/lib/core/src/application/frp.rs [hunk 14]
```diff
@@ -332,25 +339,26 @@ macro_rules! define_endpoints {
         /// Frp output setters.
         #[derive(Debug,Clone,CloneRef)]
         pub(crate) struct FrpOutputsSource {
-            $($out_field : frp::Any<$($out_field_type)*>),*
+            $(pub(crate) $out_field : enso_frp::Any<$($out_field_type)*>),*
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply src/rust/ensogl/lib/core/src/application/frp.rs]: Leaving it for now. I had some problem with macro def here. I'll add comment about it.

File: src/rust/ide/lib/span-tree/src/node.rs (hunk 5 above)
Comment (wdanilo) [reply src/rust/ide/lib/span-tree/src/node.rs:74]: Incompatible types. `self : Node<T>` , result `Node<S>`

File: src/rust/ide/lib/span-tree/src/node.rs (hunk 7 above)
Comment (wdanilo) [reply src/rust/ide/lib/span-tree/src/node.rs:146]: incompatible types (as above)

File: src/rust/ide/view/graph-editor/src/component/node/input/area.rs [hunk 15]
```diff
@@ -0,0 +1,661 @@
...
+                    node.frp.set_active           <+ parent_frp.set_active;
+                    node.frp.set_hover            <+ parent_frp.set_hover;
+                    node.frp.set_parent_connected <+ parent_frp.set_parent_connected;
+                }
+            }
+            let new_parent_frp = Some(node.frp.output.clone_ref());
+            let new_shift = if !not_a_port { 0 } else { builder.shift + node.payload.local_index };
+            builder.nested(new_parent,new_parent_frp,is_parensed,new_shift)
+        });
+    }
```
Comment (farmaazon) [inline src/rust/ide/view/graph-editor/src/component/node/input/area.rs:551]: This function is big. Can we move some logic to a separate functions?

File: src/rust/ide/view/graph-editor/src/component/node/input/area.rs [hunk 16]
```diff
@@ -0,0 +1,661 @@
...
+
+    /// Initializes FRP network for every port. Please note that the networks are connected
+    /// hierarchically (children get events from parents), so it is easier to init all networks
+    /// this way, rather than delegate it to every port.
+    fn init_port_frp_on_new_expression(&self, expression:&mut Expression) {
```
Comment (farmaazon) [inline src/rust/ide/view/graph-editor/src/component/node/input/area.rs:556]: If this is port FRP initialization, what is this network we build in build_port_shapes_on_new_expression?

File: src/rust/ensogl/lib/theme/src/lib.rs [hunk 17]
```diff
@@ -101,105 +104,101 @@ macro_rules! define_theme {
 // === Light Theme & Modules ===
 // =============================
 
-define_default_theme! { light_theme
+define_default_theme! { __light_theme__
```
Comment (farmaazon) [inline src/rust/ensogl/lib/theme/src/lib.rs:107]: Let's make it a normal name, but module private, and then reexport it in `builtin`

Same for dark theme

File: src/rust/ide/view/graph-editor/src/component/node/input/area.rs [hunk 18]
```diff
@@ -0,0 +1,661 @@
...
+                || skip_opr
+                || node.is_token()
+                || builder.parent_parensed;
+
+            if let Some(id) = node.ast_id {
+                self.model.id_crumbs_map.borrow_mut().insert(id,node.crumbs.clone_ref());
+            }
+
+            if DEBUG {
+                let indent = " ".repeat(4*builder.depth);
```
Comment (farmaazon) [inline src/rust/ide/view/graph-editor/src/component/node/input/area.rs:470]:
```suggestion
                let indent  = " ".repeat(4*builder.depth);
```

**Review by farmaazon — APPROVED** [review]

File: src/rust/ensogl/lib/core/src/display/style/path.rs [hunk 19]
```diff
@@ -4,6 +4,31 @@ use crate::prelude::*;
...
+// ==================
+// === StaticPath ===
+// ==================
+
+/// Static version of `Path`. Defined as a strongly typed wrapper over `&'static str`.
+#[derive(Clone,Copy,Debug,PartialEq,Eq)]
+#[allow(missing_docs)]
+pub struct StaticPath {
+    pub str : &'static str
+}
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply src/rust/ensogl/lib/core/src/display/style/path.rs:16]: Will be done in next refactoring.

File: src/rust/ide/view/graph-editor/src/component/node/input/area.rs [hunk 20]
```diff
@@ -0,0 +1,661 @@
...
+            let is_expected_arg       = node.is_expected_argument();
+            let span                  = node.span();
+            let mut size              = span.size.value;
+            let mut index             = span.index.value;
+            let offet_from_prev_tok   = node.offset.value - info.prev_tok_local_index;
+            info.prev_tok_local_index = node.offset.value + size;
+            viz_code += &" ".repeat(offet_from_prev_tok);
+            if node.children.is_empty() {
+                viz_code += &code[index..index+size];
+            }
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply src/rust/ide/view/graph-editor/src/component/node/input/area.rs:135]: Where would you add it in generator module? I'll add it in next PR there

File: src/rust/ide/view/src/searcher.rs [hunk 21]
```diff
@@ -197,14 +197,14 @@ impl View {
             eval frp.select_suggestion((id) model.list.select_entry(id));
             source.selected_entry <+ model.list.selected_entry;
             source.size           <+ height.value.map(|h| Vector2(SEARCHER_WIDTH,*h));
-            source.is_visible     <+ model.list.size.map(|size| size.x * size.y > std::f32::EPSILON);
+            source.is_visible     <+ model.list.size.map(|size| size.x*size.y > std::f32::EPSILON);
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply src/rust/ide/view/src/searcher.rs:200]: As agreed, let's modify style guide

File: src/rust/lib/frp/src/nodes.rs [hunk 22]
```diff
@@ -733,7 +733,12 @@ impl          IntoParam<String>    for &str { fn into_param(self) -> String {sel
 // ===============
 
 #[derive(Debug)]
-pub struct SamplerData  <Out=()> { src:Box<dyn std::any::Any>, value:RefCell<Out> }
+pub struct SamplerData  <Out=()> {
+    src        : Box<dyn std::any::Any>,
+    value      : RefCell<Out>,
+    /// Used to cache the output even if no connection is present YET.
+    self_watch : RefCell<Option<Box<dyn std::any::Any>>>
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply src/rust/lib/frp/src/nodes.rs:740]: As agreed, adding better docs there.

File: src/rust/ide/view/graph-editor/src/component/node/input/area.rs (hunk 15 above)
Comment (wdanilo) [reply src/rust/ide/view/graph-editor/src/component/node/input/area.rs:595]: In next PR Ill try to split it even more

File: src/rust/ensogl/lib/theme/src/lib.rs (hunk 17 above)
Comment (wdanilo) [reply src/rust/ensogl/lib/theme/src/lib.rs:107]: Leaving it for now, as you cannot then make `light_theme` private and reexport it in sub-omdule. Let's talk about it again,

**Pull request merged**